babel/localedata.py

```diff
@@ -89,9 +92,10 @@ def load(name, merge_inherited=True):
     :param merge_inherited: whether the inherited data should be merged into
                             the data of the requested locale
     :raise `IOError`: if no locale data file is found for the given locale
-                      identifer, or one of the locales it inherits from
+                      identifier, or one of the locales it inherits from
     """
     _cache_lock.acquire()
+    name = 'root' if name == 'root' else normalize_locale(name)
```

Comment:
The relationship between `Locale.__init__`, `Locale._data`, `localedata.exists` and `localedata.load` seemed off.

1) In `Locale.__init__`, we check if the passed in identifier is valid by calling `localedata.exists`, which normalizes the identifier to make the check (both before and after this PR). https://github.com/python-babel/babel/blob/b7dda2a0de98344426359087a3d3c239b74de548/babel/core.py#L167 If the check succeeds we return a `Locale` instance.

2) When the lazy property `Locale._data` is accessed, there is a call to `localedata.load`, the identifier passed in is not normalized. https://github.com/python-babel/babel/blob/b7dda2a0de98344426359087a3d3c239b74de548/babel/core.py#L364

This would allow the following unexpected behavior:
```
Python 2.7.6 (default, Nov 13 2018, 12:45:42)
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import babel
>>> a = babel.Locale('en')  # lowercase
>>> b = babel.Locale('EN')  # uppercase
>>> len(a._data). # just to force the lazy load
39
>>> len(b._data). # whoops!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../venv/local/lib/python2.7/site-packages/babel/core.py", line 364, in _data
    self.__data = localedata.LocaleDataDict(localedata.load(str(self)))
  File "/.../venv/local/lib/python2.7/site-packages/babel/localedata.py", line 113, in load
    with open(filename, 'rb') as fileobj:
IOError: [Errno 2] No such file or directory: '/.../venv/local/lib/python2.7/site-packages/babel/locale-data/EN.dat'
```

On case-insensitive filesystems, it's not as dramatic but still surprising.
```
Python 2.7.15 (default, Oct  2 2018, 11:47:18)
[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import babel
>>> a = babel.Locale('en')  # lowercase
>>> b = babel.Locale('EN')  # uppercase
>>> len(a._data)  # just to force the lazy load
40
>>> len(b._data)
40
>>> babel.localedata._cache.keys()  # whoops! stored the same data twice
['en', 'root', 'EN']
```

This PR attempts to remediate by normalizing the identifier here.

Reply:
Yeah, this can happen... People usually shouldn't be using the `babel.Locale()` ctor by hand anyway, but `babel.Locale.parse()` instead.

We're acutely aware that's not the best API though. :)